internal/build/resolver/resolver.go

```diff
@@ -0,0 +1,60 @@
+package resolver
+
+import (
+	"context"
+	"errors"
+
+	"github.com/buildkite/cli/v3/internal/build"
+	"github.com/buildkite/cli/v3/internal/pipeline"
+)
+
+var resolved bool
+var resolvedBuild *build.Build
+var resolvedError error
+
+// BuildResolverFn is a function for finding a build. It returns an error if an irrecoverable scenario happens and
+// should halt execution. Otherwise, if the resolver does not find a build, it should return (nil, nil) to indicate
+// this. ie. no error occurred, but no build was found either
+//
+// Some build resolvers may need to limit scope to a pipeline so that is provided to the resolver at runtime
+type BuildResolverFn func(context.Context, pipeline.Pipeline) (*build.Build, error)
+
+type AggregateResolver []BuildResolverFn
+
+// Resolve is a BuildResolverFn that wraps up a list of resolvers to loop through and try find a build. The first build
+// to be found will be returned. If none are found, it won't return an error to match the expectation of a
+// BuildResolverFn
+//
+// This is safe to call multiple times, the same result will be returned
+func (ar AggregateResolver) Resolve(ctx context.Context, p pipeline.Pipeline) (*build.Build, error) {
+	// short-circuit if this has been called before
+	if resolved {
```

Comment:
We could do something like this for the pipeline resolver as well so that its safe to call multiple times and it gives the same result.

That way, instead of a build resolver accepting a fully resolved pipeline, it could accept a PipelineResolverFn and call that to get its pipeline (if it needs one) and if its been called before it won't do all the API/IO logic again

Reply:
I think this might be a good idea. I've added it to the pipeline as well. I'll still add a test against it as well though